File: app/src/main/java/com/robotnec/budget/app/adapters/transaction/TransactionsAdapter.java
package com.robotnec.budget.app.adapters.transaction;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import com.annimon.stream.Collectors;
import com.annimon.stream.Stream;
import com.robotnec.budget.core.service.aggregation.impl.TimeSpan;
import com.robotnec.budget.core.service.aggregation.impl.TransactionAggregation;

import java.util.ArrayList;
import java.util.List;

/**
 * @author zak <zak@swingpulse.com>
 */
public class TransactionsAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    static final int VIEW_TYPE_HEADER = 1;
    static final int VIEW_TYPE_ITEM = 2;

    private final List<TransactionListItem> items;
    private LayoutInflater layoutInflater;

    public TransactionsAdapter(Context context) {
        layoutInflater = LayoutInflater.from(context);
        items = new ArrayList<>();
        setHasStableIds(true);
    }

    public void update(TransactionAggregation aggregation) {
        this.items.clear();
        this.items.addAll(toTransactionListItems(aggregation));
        notifyDataSetChanged();
    }

    private List<TransactionListItem> toTransactionListItems(TransactionAggregation aggregation) {
        List<TransactionListItem> items = new ArrayList<>();
        Stream.of(aggregation.get(TransactionAggregation.Sorting.DESC)).forEach(entry -> {
            TimeSpan span = entry.getKey();
            items.add(new HeaderItem(
                    span.getStartDate().toLocalDate(),
                    aggregation.getSum(span)));
            items.addAll(Stream.of(entry.getValue())
                    .map(TransactionItem::new)
                    .collect(Collectors.toList()));
        });
        return items;
    }

    @Override
    public long getItemId(int position) {
        return items.get(position).getId();
    }

    @Override
    public int getItemViewType(int position) {
        return items.get(position).getViewType();
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        if (viewType == VIEW_TYPE_HEADER) {
            return HeaderItem.createViewHolder(layoutInflater, parent);
        } else {
            return TransactionItem.createViewHolder(layoutInflater, parent);
        }
    }

    @Override
    public int getItemCount() {
        return items.size();
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        items.get(position).bindViewHolder(holder);
    }
}

File: app/src/main/java/com/robotnec/budget/app/adapters/transaction/TransactionsAdapter.kt
package com.robotnec.budget.app.adapters.transaction

import android.content.Context
import androidx.recyclerview.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup
import com.robotnec.budget.core.service.aggregation.impl.TransactionAggregation
import java.util.ArrayList

/**
 * @author zak zak@swingpulse.com>
 */
class TransactionsAdapter(context: Context) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    private val items: MutableList<TransactionListItem>
    private val layoutInflater: LayoutInflater = LayoutInflater.from(context)

    init {
        items = ArrayList()
        setHasStableIds(true)
    }

    fun update(aggregation: TransactionAggregation) {
        this.items.clear()
        this.items.addAll(toTransactionListItems(aggregation))
        notifyDataSetChanged()
    }

    private fun toTransactionListItems(aggregation: TransactionAggregation): List<TransactionListItem> {
        val items = mutableListOf<TransactionListItem>()

        aggregation[TransactionAggregation.Sorting.DESC]
                .forEach { (span, transactions) ->
                    items.add(HeaderItem(span.startDate.toLocalDate(), aggregation.getSum(span)))
                    items.addAll(transactions
                            .map { TransactionItem(it) }
                            .toList())
                }
        return items
    }

    override fun getItemId(position: Int): Long {
        return items[position].id
    }

    override fun getItemViewType(position: Int): Int {
        return items[position].viewType
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return if (viewType == VIEW_TYPE_HEADER) {
            HeaderItem.createViewHolder(layoutInflater, parent)
        } else {
            TransactionItem.createViewHolder(layoutInflater, parent)
        }
    }

    override fun getItemCount(): Int {
        return items.size
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        items[position].bindViewHolder(holder)
    }

    companion object {
        internal const val VIEW_TYPE_HEADER = 1
        internal const val VIEW_TYPE_ITEM = 2
    }
}
